### DISCUSSION TITLE
Porting Beehave to C#

### DISCUSSION THREAD

**Discussion by zvolchak** [bitbrain/beehave#280] [Show and tell]
Hi,

I've started porting Beehave to support C# natively for my side project (and for educational purposes). I've currently only ported the bare-minimum to get the Color Changing example working as a proof-of-concept to see if that would even work - and it did so far... so I figured maybe this port might be useful to anybody else in the Beehave community. I've seen [this "C# support" discussion](https://github.com/bitbrain/beehave/discussions/221) and thought native C# support might be more beneficial long term (just a gut feeling without any substantial arguments to support this claim).

At this point **I'm curious to get any recommendations/feedback** on the project structure, naming conventions and such, before I get too deep into it. Originally, I wanted to integrate C# code along side GD code within the same branch - which would ship both types together and allow the choice during the install, but had troubles with gdUnit4 addon to compile properly (and some other minor issues). So I've just created a separate branch and removed GD scripts to focus just on C# for now.

Here is my forked repo:
https://github.com/zvolchak/beehave-c-sharp/tree/godot-csharp-4.x

**Comment by bitbrain** [comment]
Thank you @zvolchak for your efforts. There had been some discussions on this topic already in the past and we came to a conclusion that rather than having two versions for C# and GDScript, it makes more sense to convert Beehave into a GDExtension (so it can be used by both, with one single codebase in C++)

@warriormaster12 ported Beehave to C++ already but this is inofficial yet: https://github.com/warriormaster12/beehave-cpp

My idea is to start development on Beehave V3 based on C++ alone, so all the nodes can be used both in C# and GDScript properly. That would render both the old V2 version (GDScript only) and your C# version unecessary.

If you are happy with it, we can close this discussion here and instead I'll start a fresh discussion for "Beehave V3" where we can discuss "must haves"

**Reply by warriormaster12** [reply]
@bitbrain I’m still interested to continue doing the port if I don’t have to do it alone 😅

**Reply by zvolchak** [reply]
Hm interesting. I'll tale a look into the cpp port - maybe I could contribute there instead. 
I'd like to start using some sort of behavior tree in my prototype game as soon as possible. Even something basic (like what I have in my fork currently) with just a few core nodes to get the tree running would be enough to start playing with the system. Hopefully I'll figure out GDExtension quick enough to contribute...

>  all the nodes can be used both in C# and GDScript properly.

Yes, I definitely like that approach of having just one codebase to work for both/multiple worlds.

> we can close this discussion here and instead I'll start a fresh discussion for "Beehave V3"

Yes, sounds good.

**Comment by bitbrain** [comment]
Created https://github.com/bitbrain/beehave/discussions/285 as a starting point for discussions. @zvolchak @warriormaster12 feel free to comment on there what your needs would be for Beehave V3 and we can go from there.

**Comment by shayneoneill** [comment]
@zvolchak Can you set that repo to private if your not going to continue with it. I tried to install your fork then got a bit confused when I realised there isnt actually any C# in it. I kind of think the protocol for github really should be keep things private until theres something to show, since google exposes wish-list projects as real ones

**Reply by zvolchak** [reply]
Yes fair enough - I've made it private.
